Add LICENSE to site examples

babilen5 opened the pull request:
Adding an explicit license to the example directory clarifies the
licensing of code snippets that have been included in the
documentation.

The license of these examples was previously up to interpretation as
the presence of a LICENSE file in `site/` (cf. [0]) implied CC-BY
4.0 even though licensing code under that license is recommended
against even by creative commons themselves (cf. [1]).

Including this LICENSE file makes it clear that examples are licensed
under ASLv2 which makes it easier to adapt and integrate their code
into other projects.

[0]: https://github.com/kubernetes-sigs/kind/blob/master/site/LICENSE
[1]: https://creativecommons.org/faq/#can-i-apply-a-creative-commons-license-to-software

babilen5 commented:
/assign @bentheelder

nikhita commented:
@babilen5 Looking at the license added in this PR, it appears to be CC-BY-SA-4.0 (please correct me if I'm mistaken!).

I'm also not sure what ASLv2 stands for..is it another name for Apache License v2?

babilen5 commented:
Thank you, @nikhita. I updated the license to the originally intended `Apache License v2` (to match kind's licensing).

nikhita commented:
Lgtm but IANAL though :) 

@swinslow @caniszczyk -- can you take a look at this from a legal perspective? Some background on the PR:

The presence of https://github.com/kubernetes-sigs/kind/blob/master/site/LICENSE seems to imply that the scripts in `examples/` are covered by CC-BY-SA 4.0, but licensing code under that license is recommended against by creative commons themselves (https://creativecommons.org/faq/#can-i-apply-a-creative-commons-license-to-software). This makes it hard to adapt and integrate the scripts into other projects.

Does adding a new Apache v2 LICENSE in the examples directly help with this issue?

babilen5 commented:
You can find additional information on this in [this discussion on Slack](https://kubernetes.slack.com/archives/CEKK1KTN2/p1598344359052800).

@BenTheElder clarified that Apache License, Version 2.0 was indeed the intended license for these examples and that CC-BY-SA 4.0 is intended for (textual) documentation itself.

The intention of this PR is to clarify the licensing of the included examples and I am happy to make changes to achieve that goal in a way that suits you best. It is just that I deem the licensing of these examples to be ambiguous.

Another approach would be to include a file similar to [Debian's machine readable copyright file](https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/), but I do not think that this is customarily used within the project.

neolit123 commented:
Hi. Examples aside, i am curious where is the clause that allows k/website and the kind website to use cc-by-4.0?

nikhita commented:
k/website -> https://github.com/kubernetes/website/blob/master/LICENSE (CC-BY-SA 4.0)

kind website -> https://github.com/kubernetes-sigs/kind/blob/master/site/LICENSE (CC-BY-SA 4.0)

From https://github.com/kubernetes/kubernetes-template-project#kubernetes-template-project, 

> a LICENSE which must be Apache 2.0 for code projects, or Creative Commons 4.0 for documentation repositories, without any custom content

neolit123 commented:
@nikhita 
is there are more official / umbrella CNCF statement outside of the template?
(not that i question the template contents).

nikhita commented:
>  is there are more official / umbrella CNCF statement outside of the template?

Not that I know of. Perhaps @swinslow or @caniszczyk know.

babilen5 commented:
Thank you for your comments. Let me summarise my understanding at this point.

The intention is for kind's documentation to be published under [CC-BY 4.0][cc-by-4.0] and its examples as [Apache License, Version 2.0][aslv2].

This is the current situation:

- This PR has been opened to clarify the licensing of the examples by including the [Apache License, Version 2.0][aslv2] in the directory containing the examples
- [kind's documentation][kind-doc] mentions `© 2020 The Kubernetes Authors | Documentation Distributed under CC BY 4.0` with a link to [CC-BY 4.0][cc-by-4.0]
- kind ships the same [LICENSE][kind-site-license] as [kubernetes/website][k8s-website-license], namely [cc-by-4.0]
- Examples in, for example, [kubernetes/website/content/en/examples/examples.go][k8s-website-example] contain an explicit licensing statement referencing [Apache License, Version 2.0][aslv2]
- [CC-BY-SA-4.0][cc-by-sa-4.0] appears to not be used
- The [Kubernetes template project][k8s-template-project] mentions `a LICENSE which must be Apache 2.0 for code projects, or Creative Commons 4.0 for documentation repositories, without any custom content` as pointed out by @nikhita 

[aslv2]: https://www.apache.org/licenses/LICENSE-2.0.html
[cc-by-4.0]: https://creativecommons.org/licenses/by/4.0/
[cc-by-sa-4.0]: https://creativecommons.org/licenses/by-sa/4.0/legalcode
[k8s-template-project]: https://github.com/kubernetes/kubernetes-template-project#kubernetes-template-project,
[k8s-website-example]: https://github.com/kubernetes/website/blob/master/content/en/examples/examples.go
[k8s-website-license]: https://github.com/kubernetes/website/blob/master/LICENSE
[kind-doc]: https://kind.sigs.k8s.io/
[kind-site-license]: https://github.com/kubernetes-sigs/kind/blob/master/site/LICENSE

swinslow commented:
Hi all, a few comments:

+1 on this PR, I think adding the LICENSE file here with the Apache-2.0 text makes sense.

What I'd additionally recommend would be adding, at the top of each source code file in the examples/ directory, an Apache-2.0 notice. This could be either or both:
* a copy of the standard Apache-2.0 license header, such as used [elsewhere in kind](https://github.com/kubernetes-sigs/kind/blob/master/pkg/cmd/doc.go); and/or
* an SPDX short-form license identifier, in the form: `// SPDX-License-Identifier: Apache-2.0` (changing the comment prefix characters as needed)

By putting a license notice directly in each source code file, that file's license will be clearer regardless of which subdirectory it's in.

The [CNCF charter, section 11(f)](https://github.com/cncf/foundation/blob/master/charter.md#11-ip-policy) is where you can find the CNCF IP policy regarding documentation being received and made available under CC-BY-4.0.

babilen5 commented:
Thanks, @swinslow 

> What I'd additionally recommend would be adding, at the top of each source code file in the examples/ directory, an Apache-2.0 notice

I'm happy to do this, but am mindful of the effect on readability of the documentation (assuming this will be rendered)

swinslow commented:
Thanks @babilen5. If adding the license notice via the methods I described isn't feasible b/c of how it'll be rendered, it's fine to express it in any other form (strings, comments, whatever). 

I'm just encouraging inserting into the source file itself, a statement that that file is under Apache-2.0, so that it travels along with the file when it's used elsewhere.

babilen5 commented:
From a licensing perspective including a header in every file is definitely the most explicit way and it'll ensure that files can be moved around without inadvertently affecting their licensing.

More than happy to do this. Would this be desired by kind?

BenTheElder commented:
/ok-to-test
I'm OK adding the header if necessary, but I'd just as soon update the notice at the bottom of the site and add a LICENSE file to the examples directory, if that's sufficient.

BenTheElder commented:
For clarity: the contents of these files are inlined in documentation pages as examples, I suppose it wouldn't be bad to have a license comment on each sample, but for short samples in particular it seems unusual and heavy handed when viewed on the site.

Given that we have a license notice at the bottom of each page in addition to (the now corrected) notice in the repo source tree I would hope that's sufficient.

@babilen5 let's update [site/layouts/partials/footer.html](https://github.com/kubernetes-sigs/kind/blob/master/site/layouts/partials/footer.html)

```patch
<div class="footer">
-  <p class="copyright">© {{now.Year}} The Kubernetes Authors | Documentation Distributed under <a class="nowrap" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a></p>
+  <p class="copyright">© {{now.Year}} The Kubernetes Authors | Documentation Distributed under <a class="nowrap" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a> | <a href="https://github.com/kubernetes-sigs/kind/tree/master/site/static/examples">Examples</a> Distributed under <a class="nowrap" href="https://www.apache.org/licenses/LICENSE-2.0">Apache-2.0</a></p>
  {{if .GitInfo}}<p>Last Updated on {{.GitInfo.AuthorDate}} in <span id="commit"><a href="https://github.com/kubernetes-sigs/kind/commit/{{.GitInfo.Hash}}">{{.GitInfo.AbbreviatedHash}}</a></span></p>{{end}}
</div>
```

babilen5 commented:
> I suppose it wouldn't be bad to have a license comment on each sample, but for short samples in particular it seems unusual and heavy handed when viewed on the site.

That was exactly my concern, @BenTheElder. I think for documentation, readability is important and including many lines of boilerplate would IMHO distract from the content.

A suitable middle ground might be to add something along the lines of:

```
# Copyright 2020 The Kubernetes Authors
# SPDX-License-Identifier: Apache-2.0
```

> Given that we have a license notice at the bottom of each page in addition to (the now corrected) notice in the repo source tree I would hope that's sufficient.

That would be my hope as well.

> @babilen5 let's update site/layouts/partials/footer.html

Good idea, I've included the proposed statement and added a link to the CNCF Charter, section 11 to the commit message.

BenTheElder commented:
[btw don't worry about the CI flaking here, we know about the infra issues, I'll bypass it and focus on the diff and the site preview: https://deploy-preview-1829--k8s-kind.netlify.app/]

babilen5 commented:
https://deploy-preview-1829--k8s-kind.netlify.app/ looks fine to me. What are your thoughts on adding the comments?

BenTheElder commented:
i'm going to merge this as-is for now and keep an eye out for replies from the others, if it's OK not to I'd just as soon leave at as the license file in the repo and the notice in the footer for now.

BenTheElder commented:
thank you!

babilen5 commented:
Brilliant, thanks everyone!

swinslow commented:
LGTM, thanks @babilen5 and @BenTheElder!

BenTheElder commented:
thanks for the input on this everyone!
